### PR TITLE
[WIP] HSA experiments

### PR DESCRIPTION
**Pull request by nimlgen**
Early prototype of replacing hip with hsa. Kernels run fine. Need to work on sync and less hack-ish implementation.

### CODE REVIEW

**Comment by geohot** [comment]
Nice! Adding a $700 bounty for this, assuming it can run with ~0 overhead

File: extra/hsa/demo.py [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+  dispatch_packet_addr = ctypes.addressof(base_address[index & (queue.contents.size - 1)])
+  dispatch_packet_ptr = ctypes.pointer(base_address[index & (queue.contents.size - 1)])
+  print("dispatch_packet_addr", dispatch_packet_addr)
+
+  dispatch_packet_ptr.contents.setup |= 1 << hsa.HSA_KERNEL_DISPATCH_PACKET_SETUP_DIMENSIONS
+
+  dispatch_packet_ptr.contents.workgroup_size_x = 32
+  dispatch_packet_ptr.contents.workgroup_size_y = 1
+  dispatch_packet_ptr.contents.workgroup_size_z = 1
+  dispatch_packet_ptr.contents.grid_size_x = 2 * 32 # opencl-like
```
Comment (geohot) [inline extra/hsa/demo.py:216]: Cool, wonder if it supports non divisible ones

Comment (nimlgen) [reply extra/hsa/demo.py:380]: Yeah, works fine on the simple example I ran. Seems it just masks the last lanes of the workgroup.

File: tinygrad/runtime/ops_hsa.py [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+import tinygrad.runtime.autogen.hip as hip
+from tinygrad.helpers import DEBUG, getenv, init_c_var
+from tinygrad.helpers import from_mv, round_up, to_mv, colored, init_c_struct_t
+from tinygrad.device import Compiled, LRUAllocator, MallocAllocator, BufferOptions, JITRunner, Device, Buffer, update_stats, Compiler
+from tinygrad.renderer.cstyle import HIPRenderer
+from tinygrad.codegen.kernel import LinearizerOptions
+from tinygrad.runtime.compiler.hip_comgr import compile_hip
+
+import gpuctypes.hsa as hsa
+from tinygrad.runtime.driver.hsa import *
```
Comment (geohot) [inline tinygrad/runtime/ops_hsa.py:13]: Prefer not to use import *

**Pull request closed without merging**